### ISSUE TITLE
[no-unused-vars] False positive when var is used in @jsxFrag

### ISSUE THREAD

**Issue by dbartholomae** [typescript-eslint/typescript-eslint#2514]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  extends: [
    "plugin:@typescript-eslint/recommended",
    "plugin:react/recommended"
  ],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: "tsconfig.json",
    sourceType: "module",
    ecmaFeatures: {
      jsx: true,
    }
  },
  rules: {
    "react/prop-types": 0
  }
}
```

```TSX
/* @jsx MD */
/* @jsxFrag Fragment */
import MD from "..";
import { Fragment } from ".";
<>Test</>;
```

**Expected Result**

This should lint just fine.

**Actual Result**

```
warning  'Fragment' is defined but never used  @typescript-eslint/no-unused-vars
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.1-alpha.5` |
| `@typescript-eslint/parser`        | `4.1.1-alpha.5` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `12.18.3` |

**Comment by dbartholomae** [comment]
Not sure if this will be solved by #1856 but linking the issue as it might be related

**Comment by bradzacher** [comment]
We have `parserOption`s for both the [pragma](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsjsxpragma) and the [fragment name](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsjsxfragmentname), both of which we are able to infer from the tsconfig if you're using type-aware parsing.

However this style of config will need to be built specifically - which is why support for inline config comments has not implemented.

Happy to accept a PR.

**Comment by dbartholomae** [comment]
Thanks! I'll first need to finish the project where I'm using this, but happy to work on a PR for this next. I haven't yet contributed to typescript-eslint, so a small hint where to start would be greatly appreciated (only if it doesn't take you much time). It seems like the jsx pragma is already implemented? If you think, adding the jsxFrag pragma is a good first issue, I'm happy to take it, though a hint where to start would be greatly appreciated (I would assume somewhere in scope-manager?).
This might not be the most important ticket for this library at the moment, though, so if there is another issue where it would make more sense for me to help, I would be happy to do so :)

**Comment by bradzacher** [comment]
The way it is set up is such that concerns are separated.
- `typescript-estree` is solely concerned with coordinating with TS to produce an ESTree AST.
- `scope-manager` is solely concerned with doing scope analysis using the provided options.
- `parser` is concerned with coordinating the above packages in the context of an ESLint run.

So the feature would live in `parser`.
This is where we currently prepare the scope analysis options before calling the analyser. Right now this code just fills in the blanks if you're using type-aware parsing.

https://github.com/typescript-eslint/typescript-eslint/blob/ec7449d133e1bbca76a7df4e4b325e9bda547fb1/packages/parser/src/parser.ts#L137-L171

The feature could be to iterate the comments in the AST, and then overwrite the the option that we find  config comments for. This would match TS's behaviour, as you can configure `jsxFactory` in your tsconfig and overwrite it per-file with a config comment.

**Comment by dbartholomae** [comment]
Thanks! Unfortunately I did not get to this issue over the last weeks and due to changes in priorities won't be able to take it in the near future :(
Feel free to either close or leave open, depending on your policy for this repo :)

**Comment by daryazata** [comment]
@dbartholomae 
I'm facing the same issue still, used vars are claimed as unused quite randomly, see an example below:
![Bildschirmfoto 2022-01-07 um 14 27 12](https://user-images.githubusercontent.com/44365122/148551038-a38c6ef7-79c4-477d-9785-e270d23e3bec.png)

**Comment by bradzacher** [comment]
@daryazata please file a separate issue. That is entirely unrelated to this issue.
